### DISCUSSION TITLE
Reuse Trilium Attribute Detail Widget

### DISCUSSION THREAD

**Discussion by j13055** [TriliumNext/Trilium#4216] [Q&A] [answered]
Hello Community,

i've searched the the whole internet and github, but maybe i searched in a wrong way. however. 
My Question might be very basic. 

I have written a Right Panel Widget in which all the Attributes are listed. I want to open up the Attribute-Details Widget when i click on an attribute. The same functionality when you click on an owned Attribute in the notes.

![attributeDetails_Panelpng](https://github.com/zadam/trilium/assets/18166691/52ec4f27-ddca-42ea-9a8d-d67993b382e5)

How do i reuse the widget from Trilium from inside of trilium itself?

Thanks and i hope my question is clear enough.

**Comment by zerebos** [comment, marked as answer]
The widget itself is not accessible through the API and I don't think there's any way to trigger it either without simulating clicking the original widget.

**Reply by j13055** [reply]
Thanks for the quick response. Too bad that they aren't accessable via API. I looked through the code of the Attributes Editor widget and the Attributes Detail widget and didn't find any API Call for the related notes  list. 
But I have to say that it was a long day, I might have overseen it. In the weekend I'll look into it again.

If someone else knows, if there is an API call for related notes to the attribute then  let me know. 

Thanks so far.

**Comment by zadam** [comment]
This widget is not exposed in the API contract.

**Reply by j13055** [reply]
All right. i'll try to find a workaround. Thank you for this wonderful piece of software ;) it is a shame that i didn't know about it since recently.

**Reply by j13055** [reply]
I've achieved to have an Attribute List in the right-pane, where all my Attributes created by me are listed. I added a Click-Event Handler. All of this is in a note of JS Frontend. Now i want to run the getRelatedNotes-function from Backend. how do i do this? Do I have to create a new note from type JS Backend and run it there? How do i access the results from the JS Backend note in the JS Frontend note?
If this question is not related to this topic i'll open another discussion.

**Reply by zerebos** [reply]
You should make use of the `api.runOnBackend` function as seen in the [api docs](https://zadam.github.io/trilium/frontend_api/FrontendScriptApi.html)

It generally looks like this:
```js
const myQuery = "note.title=foo";
const myResults = await api.runOnBackend((query) => {
    // inside this function "api" is now the backend api (https://zadam.github.io/trilium/backend_api/BackendScriptApi.html)
    // make sure to return so the frontend gets the data
    return api.searchForNotes(query);
}, [myQuery]);
```

Note that the query is being given as part of the array rather than used directly inside the backend function, this is necessary because after serialization and passing context, the function will not have access to local variables, so any and all local variables need to be passed in this way.

I used search as the example here as I don't think the backend api has a `getRelatedNotes` function but rather a generic search you have full control over. So if you have a custom function to achieve this, plug it in here.

**Reply by j13055** [reply]
Thanks for the quick response. I'll try to implement this and will update here soon.

**Reply by j13055** [reply]
I am very confused what i did today. However, i get an array, but thats always of length 0. I'll continiue with this next week. Thanks for your suggestion. It helped a lot to guide me into the right direction.

**Reply by zerebos** [reply]
Feel free to ask more questions, I'll help as much as I can!

**Reply by j13055** [reply]
i was off for a while and am back now to work on this widget again. Thanks for all your help. with the help of you (and chatGPT 😄) i got somewhat what i wanted. i have a list with all the Attributes i need and the correct related notes show up correctly. my problem now is that i can't get the noteId of a related note, when i click on it. i only get "undefined"

**Reply by zerebos** [reply]
I'm not exactly sure what you're referring to here without a bit of code of what you're trying to do and not able to do.

**Comment by sottey** [comment]
It is a hack, but couldn't you use javascript to grab the attributes element and fire the click event?

**Reply by j13055** [reply]
That's what i'll try to do i guess.
